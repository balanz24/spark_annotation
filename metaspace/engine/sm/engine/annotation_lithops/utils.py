from __future__ import annotations

import json
import logging
from base64 import urlsafe_b64encode
from hashlib import blake2b

logger = logging.getLogger('annotation-pipeline')


def jsonhash(obj) -> str:
    """
    Calculates a hash for a JSON-stringifiable object. Intended for compacting large sets of
    parameters into a simple key that can be used to distinctly identify a cache entry.

    The output is collision-resistant, but shouldn't be assumed to be cryptographically secure.
    In most cases a motivated adversary could figure out the original object contents easily, as
    there's no hidden key and it's unlikely there will be much variety in the objects hashed.
    """
    json_val = json.dumps(obj, sort_keys=True)
    hash_val = blake2b(json_val.encode(), digest_size=12).digest()
    return str(urlsafe_b64encode(hash_val), 'utf-8')
