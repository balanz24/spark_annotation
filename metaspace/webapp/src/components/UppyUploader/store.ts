// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from '@vue/composition-api'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach(listener => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
