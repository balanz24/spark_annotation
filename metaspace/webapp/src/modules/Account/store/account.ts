import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
