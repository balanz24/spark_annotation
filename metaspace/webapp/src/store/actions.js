import apolloClient from '../api/graphqlClient';
import {fetchOptionListsQuery} from '../api/metadata';
import {decodeParams} from '../modules/Filters';
import {computed} from "@vue/composition-api";
import store from "../store/index";
export default {

  async initFilterLists(context) {
    if(context.state.filterListsLoading || context.state.filterLists != null)
      return;

    context.commit('setFilterListsLoading');

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    });

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds)} );

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)
    context.commit('updateFilter', filter)
  },

};
