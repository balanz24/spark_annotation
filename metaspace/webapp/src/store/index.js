import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import getters from './getters.js';
import mutations from './mutations.js';
import actions from './actions.js';
import {accountModule} from '../modules/Account';

const store = new Vuex.Store({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: {visible: false},

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE'
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule
  }
})

export default store;
