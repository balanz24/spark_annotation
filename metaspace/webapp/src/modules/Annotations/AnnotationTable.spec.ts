import { mount, config as testConfig } from '@vue/test-utils'
import AnnotationTable from './AnnotationTable.vue'
import router from '../../router'
import { initMockGraphqlClient, apolloProvider } from '../../../tests/utils/mockGraphqlClient'
import store from '../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import * as FileSaver from 'file-saver'
import { merge } from 'lodash-es'
jest.mock('file-saver')
const mockFileSaver = FileSaver as jest.Mocked<typeof FileSaver>

Vue.use(Vuex)
sync(store, router)

const blobToText = (blob: Blob) => new Promise<string>((resolve, reject) => {
  const reader = new FileReader()
  reader.addEventListener('load', () => { resolve(reader.result as string) })
  reader.addEventListener('error', (e) => { reject((e as any).error) })
  reader.readAsText(blob)
})

// HACK: Prevent AnnotationTable from waiting before loading data
delete (AnnotationTable as any).options.apollo.annotations.debounce

describe('AnnotationTable', () => {
  const mockAnnotation = {
    id: '2019-02-12_15h55m06s_HMDB-v4_2018-04-09_C19H18N2O7S2_minus_H',
    sumFormula: 'C19H18N2O7S2',
    adduct: '-H',
    ion: 'C19H18N2O7S2-H+',
    msmScore: 0.657486,
    rhoSpatial: 0.686362,
    rhoSpectral: 0.959122,
    rhoChaos: 0.998756,
    fdrLevel: 0.1,
    mz: 449.0482267014823,
    dataset: {
      id: '2019-02-12_15h55m06s',
      name: 'Untreated_3_434',
      polarity: 'POSITIVE',
      group: {
        id: '20efa10e-d765-11e8-b529-8b73ec16f26e',
        name: 'Laboratory of European Molecular Biology',
      },
      groupApproved: true,
    },
    possibleCompounds: [
      { name: 'C.I. Food Red 6', information: [{ databaseId: 'HMDB0032738' }] },
    ],
    colocalizationCoeff: 0.840809,
    offSample: false,
    offSampleProb: 0.03,
  }
  const propsData = { hideColumns: ['OffSampleProb'] }

  it('should match snapshot', async() => {
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: () => ([
          mockAnnotation,
        ]),
        countAnnotations: () => 1,
      }),
    })
    const wrapper = mount(AnnotationTable, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot when loading snapshot', async() => {
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: () => ([
          mockAnnotation,
        ]),
        countAnnotations: () => 1,
      }),
    })
    router.replace({
      name: 'annotations',
      query: {
        viewId: 'xxxx',
        ds: '2019-02-12_15h55m06s',
      },
    })
    const wrapper = mount(AnnotationTable, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('should be able to export a CSV', async() => {
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: (_: any, { offset }: any) => {
          return [
            merge({}, mockAnnotation, { id: `AnnotationId_${offset + 1}`, mz: offset + 1 }),
            merge({}, mockAnnotation, {
              dataset: { groupApproved: false },
              id: `AnnotationId_${offset + 2}`,
              mz: offset + 2,
            }),
          ]
        },
        countAnnotations: () => 4,
      }),
    })
    const wrapper = mount(AnnotationTable, { store, router, apolloProvider, propsData })
    wrapper.setData({ csvChunkSize: 2 })
    await new Promise(resolve => setTimeout(resolve, 1))

    await (wrapper.vm as any).startExport()
    expect(mockFileSaver.saveAs).toBeCalled()
    const blob: Blob = mockFileSaver.saveAs.mock.calls[0][0]
    const csv = await blobToText(blob)
    const csvWithoutDateHeader = csv.replace(/# Generated at .*\n/, '')

    expect(csvWithoutDateHeader).toMatchSnapshot()
  })
})
