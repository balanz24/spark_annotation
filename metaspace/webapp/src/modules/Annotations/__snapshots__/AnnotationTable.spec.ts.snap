// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotationTable should be able to export a CSV 1`] = `
"# URL: http://localhost/annotations?viewId=xxxx&ds=2019-02-12_15h55m06s
\\"group\\",\\"datasetName\\",\\"datasetId\\",\\"formula\\",\\"adduct\\",\\"chemMod\\",\\"ion\\",\\"mz\\",\\"msm\\",\\"fdr\\",\\"rhoSpatial\\",\\"rhoSpectral\\",\\"rhoChaos\\",\\"moleculeNames\\",\\"moleculeIds\\",\\"minIntensity\\",\\"maxIntensity\\",\\"totalIntensity\\",\\"isomers\\",\\"isobars\\",\\"offSample\\",\\"rawOffSampleProb\\",\\"isobarIons\\"
\\"Laboratory of European Molecular Biology\\",\\"Untreated_3_434\\",\\"2019-02-12_15h55m06s\\",\\"C19H18N2O7S2\\",\\"M-H\\",\\"annotations.0.chemMod\\",\\"C19H18N2O7S2-H+\\",\\"1\\",\\"0.657486\\",\\"0.1\\",\\"0.686362\\",\\"0.959122\\",\\"0.998756\\",\\"C.I. Food Red 6\\",\\"HMDB0032738\\",\\"10\\",\\"0\\",\\"7.333333333333333\\",\\"1\\",\\"2\\",\\"false\\",\\"0.03\\",\\"annotations.0.isobars.0.ion, annotations.0.isobars.1.ion\\"
\\"Laboratory of European Molecular Biology\\",\\"Untreated_3_434\\",\\"2019-02-12_15h55m06s\\",\\"C19H18N2O7S2\\",\\"M-H\\",\\"annotations.1.chemMod\\",\\"C19H18N2O7S2-H+\\",\\"2\\",\\"0.657486\\",\\"0.1\\",\\"0.686362\\",\\"0.959122\\",\\"0.998756\\",\\"C.I. Food Red 6\\",\\"HMDB0032738\\",\\"10.333333333333334\\",\\"0.3333333333333333\\",\\"7.666666666666667\\",\\"1\\",\\"2\\",\\"false\\",\\"0.03\\",\\"annotations.1.isobars.0.ion, annotations.1.isobars.1.ion\\"
\\"Laboratory of European Molecular Biology\\",\\"Untreated_3_434\\",\\"2019-02-12_15h55m06s\\",\\"C19H18N2O7S2\\",\\"M-H\\",\\"annotations.0.chemMod\\",\\"C19H18N2O7S2-H+\\",\\"3\\",\\"0.657486\\",\\"0.1\\",\\"0.686362\\",\\"0.959122\\",\\"0.998756\\",\\"C.I. Food Red 6\\",\\"HMDB0032738\\",\\"10\\",\\"0\\",\\"7.333333333333333\\",\\"1\\",\\"2\\",\\"false\\",\\"0.03\\",\\"annotations.0.isobars.0.ion, annotations.0.isobars.1.ion\\"
\\"Laboratory of European Molecular Biology\\",\\"Untreated_3_434\\",\\"2019-02-12_15h55m06s\\",\\"C19H18N2O7S2\\",\\"M-H\\",\\"annotations.1.chemMod\\",\\"C19H18N2O7S2-H+\\",\\"4\\",\\"0.657486\\",\\"0.1\\",\\"0.686362\\",\\"0.959122\\",\\"0.998756\\",\\"C.I. Food Red 6\\",\\"HMDB0032738\\",\\"10.333333333333334\\",\\"0.3333333333333333\\",\\"7.666666666666667\\",\\"1\\",\\"2\\",\\"false\\",\\"0.03\\",\\"annotations.1.isobars.0.ion, annotations.1.isobars.1.ion\\"
"
`;

exports[`AnnotationTable should match snapshot 1`] = `
<div class="el-row">
  <div class="el-table el-table--fit el-table--striped el-table--border el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition el-table--mini" id="annot-table" element-loading-text="Loading results …" tabindex="1" mock-v-loading="false">
    <div class="hidden-columns">
      <div></div>
      <div></div>
      <div></div>
      <!---->
      <div></div>
      <!---->
      <!---->
      <!---->
      <!---->
      <div></div>
      <!---->
      <!---->
      <!---->
      <!---->
      <!---->
      <!---->
      <div></div>
      <!---->
    </div>
    <div class="el-table__header-wrapper">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 550px;">
        <colgroup>
          <col name="el-table_1_column_1" width="95">
          <col name="el-table_1_column_2" width="140">
          <col name="el-table_1_column_3" width="120">
          <col name="el-table_1_column_4" width="65">
          <col name="el-table_1_column_5" width="70">
          <col name="el-table_1_column_6" width="60">
        </colgroup>
        <thead class="">
          <tr class="">
            <th colspan="1" rowspan="1" class="el-table_1_column_1     is-leaf is-sortable">
              <div class="cell">Lab<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_2     is-leaf is-sortable">
              <div class="cell">Dataset<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_3     is-leaf is-sortable">
              <div class="cell">Annotation<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_4     is-leaf is-sortable">
              <div class="cell">m/z<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_5 descending    is-leaf is-sortable">
              <div class="cell">MSM<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_1_column_6   fdr-cell  is-leaf is-sortable">
              <div class="cell">FDR<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div class="el-table__body-wrapper is-scrolling-left">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 550px;">
        <colgroup>
          <col name="el-table_1_column_1" width="95">
          <col name="el-table_1_column_2" width="140">
          <col name="el-table_1_column_3" width="120">
          <col name="el-table_1_column_4" width="65">
          <col name="el-table_1_column_5" width="70">
          <col name="el-table_1_column_6" width="60">
        </colgroup>
        <tbody>
          <tr class="el-table__row fdr-10 coloc-80">
            <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
              <div class="cell">
                <div class="cell-wrapper"><span class="cell-span"><span>Laboratory of European Molecular Biology</span></span> <svg data-icon="filter" class="cell-filter-button"></svg></div>
              </div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
              <div class="cell">
                <div class="cell-wrapper"><span class="cell-span">
            Untreated_3_434
          </span> <svg data-icon="filter" class="cell-filter-button"></svg></div>
              </div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
              <div class="cell">
                <div class="cell-wrapper">
                  <mock-el-popover trigger="hover" placement="right" class="cell-span">
                    <div slot-key="default">
                      <div class="leading-5 py-2 px-3 font-normal content text-sm text-left">
                        <!---->
                        <p class="title">
                          Candidate molecule
                        </p>
                        <ul class="p-0 list-none leading-7">
                          <li>
                            C.I. Food Red 6
                          </li>
                          <!---->
                        </ul>
                        <div>
                          <p class="title"><span>2 isomeric ions were annotated</span></p>
                          <p>
                            Check the <b>Molecules</b> panel for more candidates.
                          </p>
                        </div>
                        <div>
                          <p class="title"><span>2 isobaric ions were annotated</span></p>
                          <p class="max-w-measure-1">
                            Check the <b>Molecules</b> panel to see candidate molecules from the isobaric ions,
                            and the <b>Diagnostics</b> panel to compare the isotopic images and spectra.
                          </p>
                        </div>
                      </div>
                    </div>
                    <div slot-key="reference"><span>[C<sub class="leading-none">19</sub>H<sub class="leading-none">18</sub>N<sub class="leading-none">2</sub>O<sub class="leading-none">7</sub>S<sub class="leading-none">2</sub> - H]⁺</span>
                      <!---->
                      <!---->
                    </div>
                  </mock-el-popover> <svg data-icon="filter" class="cell-filter-button"></svg>
                </div>
              </div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
              <div class="cell">
                <div class="cell-wrapper"><span class="cell-span">
            449.0482
          </span> <svg data-icon="filter" class="cell-filter-button"></svg></div>
              </div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_5  ">
              <div class="cell"><span>0.657</span></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_1_column_6  fdr-cell">
              <div class="cell"><span>10%</span></div>
            </td>
          </tr>
          <!---->
        </tbody>
      </table>
      <!---->
      <!---->
    </div>
    <!---->
    <!---->
    <!---->
    <!---->
    <div class="el-table__column-resize-proxy" style="display: none;"></div>
  </div>
  <div class="flex justify-between items-start mt-2">
    <div class="mt-1">
      <mock-el-pagination total="1" page-size="15" page-sizes="15,20,25,30" current-page="1" layout="prev,pager,next,sizes"></mock-el-pagination>
      <div id="annot-count" class="mt-2"><b>1</b> matching record
      </div>
      <div class="mt-2">
        <div class="fdr-legend-header">
          FDR levels:
        </div>
        <div class="fdr-legend fdr-5">
          5%
        </div>
        <div class="fdr-legend fdr-10">
          10%
        </div>
        <div class="fdr-legend fdr-20">
          20%
        </div>
        <div class="fdr-legend fdr-50">
          50%
        </div>
      </div>
    </div>
    <div class="flex w-full items-center justify-end flex-wrap">
      <mock-el-popover class="mt-1">
        <div slot-key="default">
          <div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>Lab</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>Dataset</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>Annotation</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Adduct</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>m/z</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Off-sample probability %</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span> ρ<sub>spatial</sub></span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span> ρ<sub>spectral</sub></span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span> ρ<sub>chaos</sub></span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>MSM</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Database</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Molecules</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Isomers</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Isobars</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Max Intensity</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Total Intensity</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>FDR</span></div>
            <div class="cursor-pointer select-none">
              <!---->
            </div>
          </div>
        </div>
        <div slot-key="reference"><button type="button" class="el-button select-btn-wrapper relative el-button--default">
            <!---->
            <!----><span><div class="el-badge new-custom-badge sm-feature-badge">
            Columns
          <transition-stub name="el-zoom-in-center"><sup class="el-badge__content el-badge__content--undefined is-fixed">New</sup></transition-stub></div> <i class="el-icon-arrow-down select-btn-icon"></i></span></button></div>
      </mock-el-popover>
      <mock-el-popover popper-class="export-pop" class="select-btn-wrapper ml-2 mt-1">
        <div slot-key="default">
          <p class="export-option">
            Annotations table
          </p>
          <p class="export-option">
            Pixel intensities
          </p>
        </div>
        <div slot-key="reference">
          <div><button type="button" class="el-button select-btn-wrapper relative el-button--default" width="146" height="42">
              <!---->
              <!----><span>
            Export to CSV
            <i class="el-icon-arrow-down select-btn-icon"></i></span></button></div>
        </div>
      </mock-el-popover>
      <div class="ml-2 mt-1"><button type="button" class="el-button full-screen-btn el-button--default">
          <!---->
          <!----><span><svg data-icon="exit_full_screen" class="full-screen-icon"></svg></span></button></div>
    </div>
  </div>
</div>
`;

exports[`AnnotationTable should match snapshot when loading snapshot 1`] = `
<div class="el-row">
  <div class="el-table el-table--fit el-table--striped el-table--border el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition el-table--mini" id="annot-table" element-loading-text="Loading results …" tabindex="1" mock-v-loading="false">
    <div class="hidden-columns">
      <!---->
      <!---->
      <div></div>
      <!---->
      <div></div>
      <!---->
      <!---->
      <!---->
      <!---->
      <div></div>
      <!---->
      <!---->
      <!---->
      <!---->
      <!---->
      <!---->
      <div></div>
      <!---->
    </div>
    <div class="el-table__header-wrapper">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 550px;">
        <colgroup>
          <col name="el-table_2_column_9" width="120">
          <col name="el-table_2_column_10" width="65">
          <col name="el-table_2_column_11" width="70">
          <col name="el-table_2_column_12" width="60">
        </colgroup>
        <thead class="">
          <tr class="">
            <th colspan="1" rowspan="1" class="el-table_2_column_9     is-leaf is-sortable">
              <div class="cell">Annotation<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_2_column_10     is-leaf is-sortable">
              <div class="cell">m/z<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_2_column_11 descending    is-leaf is-sortable">
              <div class="cell">MSM<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
            <th colspan="1" rowspan="1" class="el-table_2_column_12   fdr-cell  is-leaf is-sortable">
              <div class="cell">FDR<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div class="el-table__body-wrapper is-scrolling-left">
      <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 550px;">
        <colgroup>
          <col name="el-table_2_column_9" width="120">
          <col name="el-table_2_column_10" width="65">
          <col name="el-table_2_column_11" width="70">
          <col name="el-table_2_column_12" width="60">
        </colgroup>
        <tbody>
          <tr class="el-table__row fdr-10 coloc-80">
            <td rowspan="1" colspan="1" class="el-table_2_column_9  ">
              <div class="cell">
                <div class="cell-wrapper">
                  <mock-el-popover trigger="hover" placement="right" class="cell-span">
                    <div slot-key="default">
                      <div class="leading-5 py-2 px-3 font-normal content text-sm text-left">
                        <!---->
                        <p class="title">
                          Candidate molecule
                        </p>
                        <ul class="p-0 list-none leading-7">
                          <li>
                            C.I. Food Red 6
                          </li>
                          <!---->
                        </ul>
                        <div>
                          <p class="title"><span>2 isomeric ions were annotated</span></p>
                          <p>
                            Check the <b>Molecules</b> panel for more candidates.
                          </p>
                        </div>
                        <div>
                          <p class="title"><span>2 isobaric ions were annotated</span></p>
                          <p class="max-w-measure-1">
                            Check the <b>Molecules</b> panel to see candidate molecules from the isobaric ions,
                            and the <b>Diagnostics</b> panel to compare the isotopic images and spectra.
                          </p>
                        </div>
                      </div>
                    </div>
                    <div slot-key="reference"><span>[C<sub class="leading-none">19</sub>H<sub class="leading-none">18</sub>N<sub class="leading-none">2</sub>O<sub class="leading-none">7</sub>S<sub class="leading-none">2</sub> - H]⁺</span>
                      <!---->
                      <!---->
                    </div>
                  </mock-el-popover> <svg data-icon="filter" class="cell-filter-button"></svg>
                </div>
              </div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_2_column_10  ">
              <div class="cell">
                <div class="cell-wrapper"><span class="cell-span">
            449.0482
          </span> <svg data-icon="filter" class="cell-filter-button"></svg></div>
              </div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_2_column_11  ">
              <div class="cell"><span>0.657</span></div>
            </td>
            <td rowspan="1" colspan="1" class="el-table_2_column_12  fdr-cell">
              <div class="cell"><span>10%</span></div>
            </td>
          </tr>
          <!---->
        </tbody>
      </table>
      <!---->
      <!---->
    </div>
    <!---->
    <!---->
    <!---->
    <!---->
    <div class="el-table__column-resize-proxy" style="display: none;"></div>
  </div>
  <div class="flex justify-between items-start mt-2">
    <div class="mt-1">
      <mock-el-pagination total="1" page-size="15" page-sizes="15,20,25,30" current-page="1" layout="pager"></mock-el-pagination>
      <div id="annot-count" class="mt-2"><b>1</b> matching record
      </div>
      <div class="mt-2">
        <div class="fdr-legend-header">
          FDR levels:
        </div>
        <div class="fdr-legend fdr-5">
          5%
        </div>
        <div class="fdr-legend fdr-10">
          10%
        </div>
        <div class="fdr-legend fdr-20">
          20%
        </div>
        <div class="fdr-legend fdr-50">
          50%
        </div>
      </div>
    </div>
    <div class="flex w-full items-center justify-end flex-wrap">
      <mock-el-popover class="mt-1">
        <div slot-key="default">
          <div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Lab</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Dataset</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>Annotation</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Adduct</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>m/z</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Off-sample probability %</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span> ρ<sub>spatial</sub></span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span> ρ<sub>spectral</sub></span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span> ρ<sub>chaos</sub></span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>MSM</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Database</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Molecules</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Isomers</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Isobars</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Max Intensity</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check invisible"></i> <span>Total Intensity</span></div>
            <div class="cursor-pointer select-none"><i class="el-icon-check"></i> <span>FDR</span></div>
            <div class="cursor-pointer select-none">
              <!---->
            </div>
          </div>
        </div>
        <div slot-key="reference"><button type="button" class="el-button select-btn-wrapper relative el-button--default">
            <!---->
            <!----><span><div class="el-badge new-custom-badge sm-feature-badge">
            Columns
          <transition-stub name="el-zoom-in-center"><sup class="el-badge__content el-badge__content--undefined is-fixed">New</sup></transition-stub></div> <i class="el-icon-arrow-down select-btn-icon"></i></span></button></div>
      </mock-el-popover>
      <mock-el-popover popper-class="export-pop" class="select-btn-wrapper ml-2 mt-1">
        <div slot-key="default">
          <p class="export-option">
            Annotations table
          </p>
          <p class="export-option">
            Pixel intensities
          </p>
        </div>
        <div slot-key="reference">
          <div><button type="button" class="el-button select-btn-wrapper relative el-button--default" width="146" height="42">
              <!---->
              <!----><span>
            Export to CSV
            <i class="el-icon-arrow-down select-btn-icon"></i></span></button></div>
        </div>
      </mock-el-popover>
      <div class="ml-2 mt-1"><button type="button" class="el-button full-screen-btn el-button--default">
          <!---->
          <!----><span><svg data-icon="exit_full_screen" class="full-screen-icon"></svg></span></button></div>
    </div>
  </div>
</div>
`;
